### PR TITLE
Feature/issue 96 extract segment building logic

### LINKED ISSUE #96: Extract Segment Building Logic from SegmentBuilder
## Parent Issue

Part of #91 - Refactor SegmentBuilder state management to prevent recurring bugs

## Overview

Extract all segment building logic from SegmentBuilder into pure, testable functions. This separates business logic from React state management.

## Current State

Segment building logic is embedded in handler functions with complex conditionals:
- `handleContinueJourney` (lines 195-273): 78 lines of segment building + validation
- `handleMarkAsDestination` (lines 275-379): 104 lines including segment building
- `resumeFromLastSegment` (lines 387-419): 32 lines of state computation

## Goals

1. Extract segment building logic to pure functions
2. Extract station/line lookup utilities
3. Achieve 100% test coverage on segment builders
4. Enable easy unit testing of complex business logic

## Dependencies

**Requires:**
- Issue #95 (Extract Validation & Types) - MUST be completed first
  - Needs `types.ts` for `SegmentRequest`, `ValidationResult`
  - Needs `validation.ts` for validation functions

### Step 1: Create File Structure

Create new files:
```
frontend/src/components/routes/SegmentBuilder/
  segments.ts                 # NEW: Segment building functions
  segments.test.ts            # NEW: Unit tests
  utils.ts                    # NEW: Station/line lookup utilities
  utils.test.ts               # NEW: Unit tests
```

### Step 2: Implement Station/Line Lookup Utilities

**File: `utils.ts`**

```typescript
import { StationResponse, LineResponse } from '@/lib/api'

/**
 * Finds a station by TfL ID
 * 
 * @param tflId - Station TfL ID
 * @param stations - Array of stations
 * @returns Station if found, null otherwise
 */
export function findStationByTflId(
  tflId: string,
  stations: StationResponse[]
): StationResponse | null {
  return stations.find(s => s.tfl_id === tflId) ?? null
}

/**
 * Finds a line by TfL ID
 * 
 * @param tflId - Line TfL ID
 * @param lines - Array of lines
 * @returns Line if found, null otherwise
 */
export function findLineByTflId(
  tflId: string,
  lines: LineResponse[]
): LineResponse | null {
  return lines.find(l => l.tfl_id === tflId) ?? null
}

/**
 * Checks if two lines are the same
 * 
 * @param line1 - First line
 * @param line2 - Second line
 * @returns true if same line
 */
export function isSameLine(
  line1: LineResponse | null,
  line2: LineResponse | null
): boolean {
  if (!line1 || !line2) return false
  return line1.tfl_id === line2.tfl_id
}
```

### Step 3: Implement Segment Building Functions

**File: `segments.ts`**

```typescript
import { StationResponse, LineResponse } from '@/lib/api'
import { SegmentRequest } from './types'
import { isStationLastInRoute } from './validation'

/**
 * Parameters for building segments when continuing journey
 */
export interface BuildContinueSegmentsParams {
  currentStation: StationResponse
  selectedLine: LineResponse
  nextStation: StationResponse
  currentSegments: SegmentRequest[]
  continueLine: LineResponse
}

/**
 * Builds segments for continuing journey (not marking as destination)
 * 
 * Handles two cases:
 * 1. Continuing on same line: only adds current station if not already last
 * 2. Changing lines: adds current station (if not last) + next station on new line
 * 
 * @param params - Segment building parameters
 * @returns New array of segments
 */
export function buildSegmentsForContinue(
  params: BuildContinueSegmentsParams
): SegmentRequest[] {
  const { currentStation, selectedLine, nextStation, currentSegments, continueLine } = params
  
  const isCurrentStationLast = isStationLastInRoute(currentStation, currentSegments)
  const isChangingLines = continueLine.tfl_id !== selectedLine.tfl_id
  
  // Start with current segments or add current station if not already last
  let updatedSegments = isCurrentStationLast
    ? [...currentSegments]
    : [
        ...currentSegments,
        {
          sequence: currentSegments.length,
          station_tfl_id: currentStation.tfl_id,
          line_tfl_id: selectedLine.tfl_id,
        },
      ]
  
  // If changing lines, add next station on new line
  if (isChangingLines) {
    updatedSegments.push({
      sequence: updatedSegments.length,
      station_tfl_id: nextStation.tfl_id,
      line_tfl_id: continueLine.tfl_id,
    })
  }
  
  return updatedSegments
}

/**
 * Parameters for building segments when marking as destination
 */
export interface BuildDestinationSegmentsParams {
  currentStation: StationResponse
  selectedLine: LineResponse
  destinationStation: StationResponse
  currentSegments: SegmentRequest[]
}

/**
 * Builds segments for marking station as destination (end of route)
 * 
 * Adds current station (if not already last) and destination station with null line_id
 * 
 * @param params - Segment building parameters
 * @returns New array of segments
 */
export function buildSegmentsForDestination(
  params: BuildDestinationSegmentsParams
): SegmentRequest[] {
  const { currentStation, selectedLine, destinationStation, currentSegments } = params
  
  const isCurrentStationLast = isStationLastInRoute(currentStation, currentSegments)
  
  // Add current station if not already last
  let updatedSegments = isCurrentStationLast
    ? [...currentSegments]
    : [
        ...currentSegments,
        {
          sequence: currentSegments.length,
          station_tfl_id: currentStation.tfl_id,
          line_tfl_id: selectedLine.tfl_id,
        },
      ]
  
  // Add destination marker (null line_id indicates destination)
  updatedSegments.push({
    sequence: updatedSegments.length,
    station_tfl_id: destinationStation.tfl_id,
    line_tfl_id: null,
  })
  
  return updatedSegments
}

/**
 * Result of computing resume state from segments
 */
export interface ResumeStateResult {
  station: StationResponse | null
  line: LineResponse | null
}

/**
 * Computes the station and line to resume from based on last segment
 * 
 * Used when editing a route or resuming after deletion
 * 
 * @param segments - Current route segments
 * @param stations - All available stations
 * @param lines - All available lines
 * @returns Resume state (station and line) or null if no segments
 */
export function computeResumeState(
  segments: SegmentRequest[],
  stations: StationResponse[],
  lines: LineResponse[]
): ResumeStateResult | null {
  if (segments.length === 0) return null
  
  const lastSegment = segments[segments.length - 1]
  
  // Find the station for the last segment
  const lastStation = stations.find(s => s.tfl_id === lastSegment.station_tfl_id) ?? null
  
  // If last segment has a line, use it; otherwise look at previous segment
  let resumeLine: LineResponse | null = null
  if (lastSegment.line_tfl_id) {
    resumeLine = lines.find(l => l.tfl_id === lastSegment.line_tfl_id) ?? null
  } else if (segments.length > 1) {
    const prevSegment = segments[segments.length - 2]
    if (prevSegment.line_tfl_id) {
      resumeLine = lines.find(l => l.tfl_id === prevSegment.line_tfl_id) ?? null
    }
  }
  
  return {
    station: lastStation,
    line: resumeLine,
  }
}

/**
 * Removes segments from a sequence number onwards and resequences remaining
 * 
 * @param sequence - Sequence number to delete from
 * @param segments - Current route segments
 * @returns New array of segments with updated sequence numbers
 */
export function deleteSegmentAndResequence(
  sequence: number,
  segments: SegmentRequest[]
): SegmentRequest[] {
  return segments
    .filter(seg => seg.sequence < sequence)
    .map((seg, index) => ({ ...seg, sequence: index }))
}
```

### Step 4: Comprehensive Unit Tests

**File: `segments.test.ts`**

Test coverage requirements:
- All segment building functions: 100% statement, 100% branch coverage
- Edge cases: empty segments, line changes, destination markers
- Complex scenarios: multiple interchanges, editing, deletion

Example test structure:
```typescript
import { describe, it, expect } from 'vitest'
import {
  buildSegmentsForContinue,
  buildSegmentsForDestination,
  computeResumeState,
  deleteSegmentAndResequence
} from './segments'

describe('buildSegmentsForContinue', () => {
  const mockStation1 = { tfl_id: '123', name: 'Station 1' } as StationResponse
  const mockStation2 = { tfl_id: '456', name: 'Station 2' } as StationResponse
  const mockLineNorthern = { tfl_id: 'northern', name: 'Northern' } as LineResponse
  const mockLineVictoria = { tfl_id: 'victoria', name: 'Victoria' } as LineResponse
  
  it('should add current station when not already last', () => {
    const result = buildSegmentsForContinue({
      currentStation: mockStation1,
      selectedLine: mockLineNorthern,
      nextStation: mockStation2,
      currentSegments: [],
      continueLine: mockLineNorthern
    })
    
    expect(result).toHaveLength(1)
    expect(result[0]).toEqual({
      sequence: 0,
      station_tfl_id: '123',
      line_tfl_id: 'northern'
    })
  })
  
  it('should not duplicate current station if already last', () => {
    const existingSegments = [{
      sequence: 0,
      station_tfl_id: '123',
      line_tfl_id: 'northern'
    }]
    
    const result = buildSegmentsForContinue({
      currentStation: mockStation1,
      selectedLine: mockLineNorthern,
      nextStation: mockStation2,
      currentSegments: existingSegments,
      continueLine: mockLineNorthern
    })
    
    expect(result).toHaveLength(1)
    expect(result[0].station_tfl_id).toBe('123')
  })
  
  it('should add next station when changing lines', () => {
    const existingSegments = [{
      sequence: 0,
      station_tfl_id: '123',
      line_tfl_id: 'northern'
    }]
    
    const result = buildSegmentsForContinue({
      currentStation: mockStation1,
      selectedLine: mockLineNorthern,
      nextStation: mockStation2,
      currentSegments: existingSegments,
      continueLine: mockLineVictoria  // Changing lines!
    })
    
    expect(result).toHaveLength(2)
    expect(result[1]).toEqual({
      sequence: 1,
      station_tfl_id: '456',
      line_tfl_id: 'victoria'
    })
  })
  
  it('should not add next station when continuing same line', () => {
    const existingSegments = [{
      sequence: 0,
      station_tfl_id: '123',
      line_tfl_id: 'northern'
    }]
    
    const result = buildSegmentsForContinue({
      currentStation: mockStation1,
      selectedLine: mockLineNorthern,
      nextStation: mockStation2,
      currentSegments: existingSegments,
      continueLine: mockLineNorthern  // Same line
    })
    
    expect(result).toHaveLength(1)
  })
})

describe('buildSegmentsForDestination', () => {
  // Similar comprehensive tests...
})

describe('computeResumeState', () => {
  // Test empty segments, single segment, multiple segments, null line_id
})

describe('deleteSegmentAndResequence', () => {
  // Test deletion at various positions, resequencing
})
```

**File: `utils.test.ts`**

```typescript
import { describe, it, expect } from 'vitest'
import { findStationByTflId, findLineByTflId, isSameLine } from './utils'

describe('findStationByTflId', () => {
  it('should find station by tfl_id', () => {
    const stations = [
      { tfl_id: '123', name: 'Station 1' } as StationResponse,
      { tfl_id: '456', name: 'Station 2' } as StationResponse
    ]
    const result = findStationByTflId('123', stations)
    expect(result?.name).toBe('Station 1')
  })
  
  it('should return null if not found', () => {
    const result = findStationByTflId('999', [])
    expect(result).toBeNull()
  })
})

// Similar tests for other utils
```

### Step 5: Refactor SegmentBuilder to Use Segment Builders

Update SegmentBuilder.tsx to import and use segment building functions:

**Before** (lines 213-234 in `handleContinueJourney`):
```typescript
const isCurrentStationLast = lastSegment?.station_tfl_id === currentStation.tfl_id
const isChangingLines = continueLine.tfl_id !== selectedLine.tfl_id

let updatedSegments = isCurrentStationLast
  ? [...localSegments]
  : [...localSegments, { sequence: localSegments.length, ... }]

if (isChangingLines) {
  updatedSegments.push({ sequence: updatedSegments.length, ... })
}
```

**After**:
```typescript
const updatedSegments = buildSegmentsForContinue({
  currentStation,
  selectedLine,
  nextStation,
  currentSegments: localSegments,
  continueLine
})
```

Apply this pattern to all segment building locations.

### Unit Tests (NEW)
- [ ] `buildSegmentsForContinue`: 8-10 test cases
  - Add current station when not last
  - Don't duplicate current station if last
  - Add next station when changing lines
  - Don't add next station when same line
  - Complex multi-segment scenarios

- [ ] `buildSegmentsForDestination`: 5-6 test cases
  - Add current + destination
  - Don't duplicate current if last
  - Destination has null line_id
  - Sequence numbering correct

- [ ] `computeResumeState`: 6-8 test cases
  - Empty segments
  - Single segment with line
  - Single segment without line
  - Multiple segments
  - Last segment is destination (null line_id)

- [ ] `deleteSegmentAndResequence`: 4-5 test cases
  - Delete from middle
  - Delete from start
  - Sequence numbers updated correctly

- [ ] Utils: 6-8 test cases for lookup functions

### Integration Tests (EXISTING)
- [ ] All 22+ existing SegmentBuilder tests must pass
- [ ] No behavior changes (logic identical, just moved)

### Coverage Goals
- [ ] Segment building functions: 100% statement, 100% branch coverage
- [ ] Utils functions: 100% coverage
- [ ] SegmentBuilder.tsx: No decrease in coverage

## Acceptance Criteria

- [ ] `segments.ts` created with all segment building functions
- [ ] `utils.ts` created with lookup utilities
- [ ] `segments.test.ts` created with 100% coverage
- [ ] `utils.test.ts` created with 100% coverage
- [ ] All functions have JSDoc comments
- [ ] SegmentBuilder.tsx refactored to use builders (no duplication)
- [ ] All 22+ existing tests pass
- [ ] No behavior changes (only refactoring)
- [ ] Pre-commit hooks pass
- [ ] TypeScript strict mode compliant
- [ ] ESLint compliant

## File Locations

**New files:**
- `frontend/src/components/routes/SegmentBuilder/segments.ts`
- `frontend/src/components/routes/SegmentBuilder/segments.test.ts`
- `frontend/src/components/routes/SegmentBuilder/utils.ts`
- `frontend/src/components/routes/SegmentBuilder/utils.test.ts`

**Modified files:**
- `frontend/src/components/routes/SegmentBuilder/SegmentBuilder.tsx`

## Estimated Effort

- Create utils.ts + tests: 1-1.5 hours
- Create segments.ts: 2-3 hours
- Create segments.test.ts: 2-3 hours
- Refactor SegmentBuilder.tsx: 1-1.5 hours
- Testing & verification: 0.5-1 hour

**Total: 5-7 hours**

## Blocked By

- #95 - Extract Validation & Types Module (MUST be completed first)

## Blocks

- #96 - Extract State Machine Logic (needs segment builders)
- #97 - Integrate Pure Functions (needs all modules)

## Success Metrics

- **Before**: Business logic embedded in handlers, hard to test
- **After**: 100% test coverage on segment builders, reusable functions

## References

- Parent issue: #91
- Required: #95
- Original SegmentBuilder: `frontend/src/components/routes/SegmentBuilder/SegmentBuilder.tsx`

---

**Status:** Ready for Implementation (after #95 complete)
**Assignee:** TBD (can be executed by any agent)

### PR DESCRIPTION
**Pull request by mnbf9rca**
resolves #96

## Summary by Sourcery

Extract segment-building and utility logic into pure modules and refactor the SegmentBuilder component to use them, improving maintainability and testability.

Enhancements:
- Extract segment building logic from SegmentBuilder into standalone pure functions (buildSegmentsForContinue, buildSegmentsForDestination, computeResumeState, deleteSegmentAndResequence)
- Move station and line lookup helpers into a dedicated utils.ts module
- Simplify SegmentBuilder component by delegating segment creation, deletion, and resume state computation to the new pure functions

Build:
- Add test-runner.sh script to streamline frontend test execution

Tests:
- Add unit tests for segment building functions
- Add unit tests for utility functions

### CODE REVIEW

File: frontend/src/components/routes/SegmentBuilder/SegmentBuilder.tsx [hunk 1]
```diff
@@ -208,61 +215,35 @@ export function SegmentBuilder({
...
-      // Add segment for current station with selected line
-      const newSegment: SegmentRequest = {
-        sequence: localSegments.length,
-        station_tfl_id: currentStation.tfl_id,
-        line_tfl_id: selectedLine.tfl_id,
-      }
-
-      updatedSegments = [...localSegments, newSegment]
+    // Check max segments limit before building
+    const maxSegmentsValidation = validateMaxSegments(localSegments.length, additionalSegments)
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply frontend/src/components/routes/SegmentBuilder/SegmentBuilder.tsx:234]: Current Behavior:
The code correctly calculates additionalSegments on lines 219-220:
const isChangingLines = !isSameLine(line, selectedLine)
const additionalSegments = isChangingLines ? 2 : 1

Why This Is Correct:
- When staying on same line: adds 1 segment (current station, if not already last)
- When changing lines: adds 2 segments (current station if not last, PLUS next station on new line)
- The buildSegmentsForContinue function (lines 80-111 in segments.ts) handles the "current station is last" case internally by checking isCurrentStationLast and NOT duplicating it
- Tests confirm this behavior: see test on line 99-118 where current station IS last and only 1 segment results

Why The Review Comment Is Wrong:
The reviewer assumes additionalSegments doesn't account for when current station is last. But this is handled INSIDE buildSegmentsForContinue, not by changing the validation count. The validation correctly checks the MAXIMUM that COULD be
added, and the builder function adds the ACTUAL amount needed.

Correct Validation Logic:
- If not changing lines: maximum 1 segment could be added
- If changing lines: maximum 2 segments could be added
- The actual implementation may add fewer if current station is already last, but validating against the maximum is correct

File: frontend/src/components/routes/SegmentBuilder/SegmentBuilder.tsx [hunk 2]
```diff
@@ -208,61 +215,35 @@ export function SegmentBuilder({
...
     }
 
-    // If user is switching to a different line, also add the nextStation segment
-    // This shows the station where the line change occurs
-    if (line.id !== selectedLine.id) {
-      // Check if nextStation is a duplicate
-      const nextValidation = validateNotDuplicate(nextStation, updatedSegments)
+    // If changing lines, validate nextStation is not a duplicate
+    if (isChangingLines) {
+      const nextValidation = validateNotDuplicate(nextStation, localSegments)
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply frontend/src/components/routes/SegmentBuilder/SegmentBuilder.tsx:242]: Current Behavior:
if (isChangingLines) {
 const nextValidation = validateNotDuplicate(nextStation, localSegments)

Why This Is Correct:
At this point in the code (BEFORE calling buildSegmentsForContinue), updatedSegments doesn't exist yet! The segments are only built on line 239. The validation happens BEFORE building to ensure we don't build invalid segments.

Additionally:
- currentStation is already validated with allowLast: true on line 212
- This validates nextStation against the existing route
- After building, nextStation will be added to the route
- Validating against localSegments (the current state) is precisely correct

Why The Review Comment Is Wrong:
The reviewer suggests using updatedSegments which doesn't exist until line 239. The validation must happen against the current state before building new segments.

File: frontend/src/components/routes/SegmentBuilder/segments.ts [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+  let resumeLine: LineResponse | null = null
+  if (lastSegment.line_tfl_id) {
+    resumeLine = lines.find((l) => l.tfl_id === lastSegment.line_tfl_id) ?? null
+  } else if (segments.length > 1) {
+    // Last segment is destination marker (null line), use previous segment's line
+    const prevSegment = segments[segments.length - 2]
+    if (prevSegment.line_tfl_id) {
+      resumeLine = lines.find((l) => l.tfl_id === prevSegment.line_tfl_id) ?? null
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply frontend/src/components/routes/SegmentBuilder/segments.ts:275]: Current Behavior:
if (lastSegment.line_tfl_id) {
 resumeLine = lines.find((l) => l.tfl_id === lastSegment.line_tfl_id) ?? null
}

Why This Is Correct:
Looking at the data model throughout the codebase:
- line_tfl_id is typed as string | null in the SegmentRequest interface
- The destination marker explicitly uses null (never empty string)
- TfL line IDs are always non-empty strings like 'northern', 'victoria', etc.
- Tests confirm this: line 191 comment explicitly states "null line_id indicates destination"
- Empty string would never be a valid TfL line ID

Why The Review Comment Is Wrong:
The reviewer assumes line_tfl_id could be an empty string or other falsy value. But:
1. TypeScript type is string | null (not string | null | undefined | "")
2. The comment on line 191 explicitly documents this: "Add destination marker (null line_id indicates destination)"
3. All tests use null for destinations, never empty strings
4. TfL API never returns empty strings for line IDs

The current truthy check is idiomatic and correct for this domain.

do you agree?
